### PR TITLE
Implement  Routing Metadata Extension

### PR DESCRIPTION
**Pull request by linux-china**
Routing Metadata Extension: 
 https://github.com/rsocket/rsocket/blob/master/Extensions/Routing.md

### CODE REVIEW

File: rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  private ByteBuf content;
+
+  public TaggingMetadata(String type, ByteBuf content) {
+    this.type = type;
+    this.content = content;
+  }
+
+  public TaggingMetadata(String type, Collection<String> tags) {
+    this.type = type;
+    CompositeByteBuf compositeByteBuf = Unpooled.compositeBuffer();
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java]: `CompositeMetadataFlyweight` allows any `ByteBufAllocator` to be passed in while here it is fixed.

File: rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    tags.stream()
+        .map(tag -> tag.getBytes(StandardCharsets.UTF_8))
+        .filter(bytes -> bytes.length > 0 && bytes.length < TAG_LENGTH_MAX)
+        .forEach(
+            bytes -> {
+              int capacity = bytes.length + 1;
+              ByteBuf tagByteBuf = buffer(capacity, capacity);
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java]: Also no option to use alternative `ByteBufAllocator` here.

File: rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        if (tagLength > 0) {
+          byte[] tagBytes = new byte[tagLength];
+          content.readBytes(tagBytes);
+          return new String(tagBytes, StandardCharsets.UTF_8);
+        } else {
+          return "";
+        }
+      }
+    };
+  }
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/metadata/TaggingMetadata.java]: A test that proves `iterator()` can be used more than once would be useful. As far as I can see that would require a `ByteBuf` slice for independent iteration.

**Comment by linux-china** [comment]
@robertroeser,  you are correct  and we should use flyweight pattern to implement this feature.  I close this PR and take a look.  If somebody can implement this feature, please implement it first.  I close this PR now and take a look at flyweight pattern.  Thanks for you to review my code.  :)

**Pull request closed without merging**